phoenix-core/src/main/java/org/apache/phoenix/expression/function/SignFunction.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.function;
+
+import java.math.BigDecimal;
+import java.sql.SQLException;
+import java.util.List;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.expression.Expression;
+import org.apache.phoenix.parse.FunctionParseNode.Argument;
+import org.apache.phoenix.parse.FunctionParseNode.BuiltInFunction;
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.PDataType;
+import org.apache.phoenix.schema.types.PDecimal;
+import org.apache.phoenix.schema.types.PNumericType;
+import org.apache.phoenix.schema.types.PRealNumber;
+import org.apache.phoenix.schema.types.PWholeNumber;
+
+/**
+ * Base class for built-in SIGN function.
+ * @since 4.3.0
+ */
+@BuiltInFunction(name = SignFunction.NAME,
+        args = { @Argument(allowedTypes = { PNumericType.class }) })
+public class SignFunction extends ScalarFunction {
+
+    public static final String NAME = "SIGN";
+
+    public SignFunction() {
+    }
+
+    public SignFunction(List<Expression> children) throws SQLException {
+        super(children);
+    }
+
+    @Override
+    public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+        Expression childExpr = children.get(0);
+        PDataType dataType = childExpr.getDataType();
+        if (childExpr.evaluate(tuple, ptr)) {
+            byte[] retBytes;
+            if (dataType instanceof PWholeNumber) {
```

Comment:
1) According to [1], signum's return type should be as same as argument type.
2) Math#signum has a special case: it will return Nan, if the argument is Nan.

So I choose to return different types for whole and real number.

When getting the result, we don't know its types. I come up with two ways to construct the object,
1) try each type one by one, or
2) achieve as it is.

[1] http://www.postgresql.org/docs/current/static/functions-math.html